### PR TITLE
LibWeb: Implement window.open

### PR DESCRIPTION
**Pull request by IdanHo**
This implementation is some-what complete, with the most common missing/broken feature being opening pages in new tabs using the "_blank" target.

This is currently broken due to 2 reasons:
 - We currently always claim the Window does not have transient activation, as we do not track the transient activation timestamp yet. This means that all such window.open calls are detected as pop-ups, and as such they are blocked. This can be easily bypassed by unchecking the 'Block Pop-ups' checkbox in the debug menu.
 - There is currently no mechanism for the WebContent process to request a new tab to be created by the Browser process, and as such the call to BrowsingContext::choose_a_browsing_context does not actually open another tab.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.h [hunk 1]
```diff
@@ -164,7 +164,18 @@ class BrowsingContext final
...
 
-    BrowsingContext* choose_a_browsing_context(StringView name, bool noopener);
+    enum class WindowType {
+        ExistingOrNone,
+        NewAndUnrestricted,
+        NewWithNoOpener,
+    };
+
+    struct ChosenBrowsingContext {
+        BrowsingContext* browsing_context;
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.h:174]: In general I think we should just always initialize plain types:
```c++
BrowsingContext* browsing_context { nullptr };
```

File: Userland/Libraries/LibWeb/HTML/Scripting/Environments.cpp [hunk 2]
```diff
@@ -151,6 +151,22 @@ void EnvironmentSettingsObject::prepare_to_run_callback()
         context->skip_when_determining_incumbent_counter++;
 }
 
+// https://html.spec.whatwg.org/multipage/urls-and-fetching.html#parse-a-url
+AK::URL EnvironmentSettingsObject::parse_url(StringView url)
+{
+    // 1. Let encoding be document's character encoding, if document was given, and environment settings object's API URL character encoding otherwise.
+    auto encoding = api_url_character_encoding();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Scripting/Environments.cpp:158]: Unused?

File: Userland/Libraries/LibWeb/HTML/Scripting/Environments.h [hunk 3]
```diff
@@ -85,6 +85,9 @@ struct EnvironmentSettingsObject
     // https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability
     virtual CanUseCrossOriginIsolatedAPIs cross_origin_isolated_capability() = 0;
 
+    // https://html.spec.whatwg.org/multipage/urls-and-fetching.html#parse-a-url
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Scripting/Environments.h:88]: Doesn't need a link here

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.h [hunk 4]
```diff
@@ -124,6 +124,11 @@ class BrowsingContext final
     HTML::WindowProxy* window_proxy();
     HTML::WindowProxy const* window_proxy() const;
 
+    BrowsingContext* opener_browsing_context() { return m_opener_browsing_context; }
+    BrowsingContext const* opener_browsing_context() const { return m_opener_browsing_context; }
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.h:128]: No more returning raw pointers pls

File: Userland/Libraries/LibWeb/HTML/CrossOrigin/Reporting.cpp [hunk 5]
```diff
@@ -26,12 +26,12 @@ void check_if_access_between_two_browsing_contexts_should_be_reported(BrowsingCo
     // 4. Let accessorAccessedRelationship be a new accessor-accessed relationship with value none.
     auto accessor_accessed_relationship = AccessorAccessedRelationship::None;
 
-    // FIXME: 5. If accessed's top-level browsing context's opener browsing context is accessor or an ancestor of accessor, then set accessorAccessedRelationship to accessor is opener.
-    if (false)
+    // 5. If accessed's top-level browsing context's opener browsing context is accessor or an ancestor of accessor, then set accessorAccessedRelationship to accessor is opener.
+    if (accessed.top_level_browsing_context().opener_browsing_context() && (accessed.top_level_browsing_context().opener_browsing_context() == &accessor || accessed.top_level_browsing_context().opener_browsing_context()->is_ancestor_of(accessor)))
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/CrossOrigin/Reporting.cpp:30]: I _knew_ `if (false)` would be helpful eventually haha

Comment (linusg) [reply Userland/Libraries/LibWeb/HTML/CrossOrigin/Reporting.cpp:30]: but can we turn this into

```cpp
auto opener_browsing_context = accessed.top_level_browsing_context().opener_browsing_context();
if (opener_browsing_context && opener_browsing_context == ...)
```

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 6]
```diff
@@ -297,6 +298,134 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
...
+            TRY(Bindings::throw_dom_exception_if_needed(vm, [&] { return target_browsing_context->navigate(request, *source_browsing_context, true); }));
+        }
+
+        // 2. If noopener is false, then set target browsing context's opener browsing context to source browsing context.
+        if (!no_opener)
+            target_browsing_context->set_opener_browsing_context(source_browsing_context);
+    }
+
+    // 13. If noopener is true or windowType is "new with no opener", then return null.
+    if (no_opener && window_type == BrowsingContext::WindowType::NewWithNoOpener)
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:422]: This should be `||` instead, no?

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 7]
```diff
@@ -693,76 +695,101 @@ BrowsingContext* BrowsingContext::choose_a_browsing_context(StringView name, boo
...
-
-        //     3. If noopener is true, then set chosen to the result of creating
-        //     a new top-level browsing context.
+        else if (true) { // FIXME: When is this the case?
+            // 1. Set windowType to "new and unrestricted".
+            window_type = WindowType::NewAndUnrestricted;
+
+            // 2. If current's top-level browsing context's active document's
+            // cross-origin opener policy's value is "same-origin" or
+            // "same-origin-plus-COEP", then:
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:731]: Align wrapped lines with step number please

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 8]
```diff
@@ -693,76 +695,101 @@ BrowsingContext* BrowsingContext::choose_a_browsing_context(StringView name, boo
...
+            if (top_level_browsing_context().active_document()->cross_origin_opener_policy().value == CrossOriginOpenerPolicyValue::SameOrigin || top_level_browsing_context().active_document()->cross_origin_opener_policy().value == CrossOriginOpenerPolicyValue::SameOriginPlusCOEP) {
+                // 2.1. Let currentDocument be current's active document.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:733]: not your fault, but no number nesting please

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp (hunk 7 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:731]: Also, we currently go up to ~120 characters before wrapping, this seems overly aggressive (same below)

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.h (hunk 1 above)
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.h:174]: (Nonnull)GCPtr please

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 9]
```diff
@@ -120,6 +122,181 @@ CSS::Screen& Window::screen()
...
+    // 1. If value is the empty string, then return true.
+    if (value.is_empty())
+        return true;
+
+    // 2. If value is "yes", then return true.
+    if (value == "yes"sv)
+        return true;
+
+    // 3. If value is "true", then return true.
+    if (value == "true")
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:224]: sv

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 10]
```diff
@@ -120,6 +122,181 @@ CSS::Screen& Window::screen()
...
+
+    // 12. Let status be the result of checking if a window feature is set, given tokenizedFeatures, "status", and false.
+    auto status = check_if_a_window_feature_is_set("status"sv, false);
+
+    // 13. If status is false, then return true.
+    if (!status)
+        return true;
+
+    // 14. Return false.
+    return true;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:298]: :thinking:

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 11]
```diff
@@ -297,6 +298,134 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
...
+        // NOTE: While this is not implemented yet, all of observable actions taken by this operation are optional (implementation-defined).
+
+        // 3. Let urlRecord be the URL record about:blank.
+        auto url_record = AK::URL("about:blank"sv);
+
+        // 4. If url is not the empty string, then parse url relative to the entry settings object, and set urlRecord to the resulting URL record, if any. If the parse a URL algorithm failed, then throw a "SyntaxError" DOMException.
+        if (!url.is_empty()) {
+            url_record = entry_settings_object().parse_url(url);
+            if (!url_record.is_valid())
+                return throw_completion(WebIDL::SecurityError::create(realm(), "url is not valid"));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:372]: `url` -> `URL`, `SecurityError` -> `SyntaxError`

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 12]
```diff
@@ -297,6 +298,134 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
...
+    else {
+        // 1. If url is not the empty string, then:
+        if (!url.is_empty()) {
+            // 1. Let urlRecord be the URL record about:blank.
+            auto url_record = AK::URL("about:blank"sv);
+
+            // 2. Parse url relative to the entry settings object, and set urlRecord to the resulting URL record, if any. If the parse a URL algorithm failed, then throw a "SyntaxError" DOMException.
+            url_record = entry_settings_object().parse_url(url);
+            if (!url_record.is_valid())
+                return throw_completion(WebIDL::SecurityError::create(realm(), "url is not valid"));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:402]: `url` -> `URL`, `SecurityError` -> `SyntaxError`

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 13]
```diff
@@ -297,6 +298,134 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
     return true;
 }
 
+// FIXME: This is based on the old 'browsing context' concept, which was replaced with 'navigable'
+// https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps
+JS::ThrowCompletionOr<HTML::WindowProxy*> Window::open_impl(StringView url, StringView target, StringView features)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:303]: Any reason why this can't be a `WebIDL::ExceptionOr`?

Comment (linusg) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:303]: (I'd prefer to convert to `ThrowCompletionOr` as late as possible, i.e. at the call site)

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 14]
```diff
@@ -499,6 +499,17 @@ WebIDL::ExceptionOr<Document*> Document::open(String const&, String const&)
     return this;
 }
 
+// https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-document-open-window
+JS::ThrowCompletionOr<HTML::WindowProxy*> Document::open(String const& url, String const& name, String const& features)
+{
+    // 1. If this is not fully active, then throw an "InvalidAccessError" DOMException exception.
+    if (!is_fully_active())
+        return throw_completion(WebIDL::SecurityError::create(realm(), "Cannot perform open on a document that isn't fully active."sv));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:507]: `InvalidAccessError`

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 6 above)
Comment (IdanHo) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:422]: Nice catch, fixed

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 10 above)
Comment (IdanHo) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:298]: Looks like I was a bit tired :sweat_smile:

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 13 above)
Comment (IdanHo) [reply Userland/Libraries/LibWeb/HTML/Window.cpp]: The only other impl method that could throw exceptions was document_tree_child_browsing_context_count, and that one used ThrowCompletionOr, so I did as well. But you're right, it's nicer this way, fixed.

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.h [hunk 15]
```diff
@@ -124,6 +124,10 @@ class BrowsingContext final
     HTML::WindowProxy* window_proxy();
     HTML::WindowProxy const* window_proxy() const;
 
+    JS::GCPtr<BrowsingContext> opener_browsing_context() const { return m_opener_browsing_context; }
+
+    void set_opener_browsing_context(BrowsingContext* browsing_context) { m_opener_browsing_context = browsing_context; }
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.h:129]: I'd argue we should use a `GCPtr` here as well :)

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 16]
```diff
@@ -297,6 +298,131 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
     return false;
 }
 
+// FIXME: This is based on the old 'browsing context' concept, which was replaced with 'navigable'
+// https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps
+WebIDL::ExceptionOr<HTML::WindowProxy*> Window::open_impl(StringView url, StringView target, StringView features)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:303]: Raw pointer -> `GCPtr` here as well

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 17]
```diff
@@ -297,6 +298,131 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
...
+        // NOTE: While this is not implemented yet, all of observable actions taken by this operation are optional (implementation-defined).
+
+        // 3. Let urlRecord be the URL record about:blank.
+        auto url_record = AK::URL("about:blank"sv);
+
+        // 4. If url is not the empty string, then parse url relative to the entry settings object, and set urlRecord to the resulting URL record, if any. If the parse a URL algorithm failed, then throw a "SyntaxError" DOMException.
+        if (!url.is_empty()) {
+            url_record = entry_settings_object().parse_url(url);
+            if (!url_record.is_valid())
+                return throw_completion(WebIDL::SyntaxError::create(realm(), "URL is not valid"));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:369]: `throw_completion()` can go now

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 18]
```diff
@@ -297,6 +298,131 @@ static bool check_if_a_popup_window_is_requested(OrderedHashMap<String, String>
...
+    else {
+        // 1. If url is not the empty string, then:
+        if (!url.is_empty()) {
+            // 1. Let urlRecord be the URL record about:blank.
+            auto url_record = AK::URL("about:blank"sv);
+
+            // 2. Parse url relative to the entry settings object, and set urlRecord to the resulting URL record, if any. If the parse a URL algorithm failed, then throw a "SyntaxError" DOMException.
+            url_record = entry_settings_object().parse_url(url);
+            if (!url_record.is_valid())
+                return throw_completion(WebIDL::SyntaxError::create(realm(), "URL is not valid"));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:399]: Likewise

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 19]
```diff
@@ -499,6 +499,17 @@ WebIDL::ExceptionOr<Document*> Document::open(String const&, String const&)
     return this;
 }
 
+// https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-document-open-window
+WebIDL::ExceptionOr<HTML::WindowProxy*> Document::open(String const& url, String const& name, String const& features)
+{
+    // 1. If this is not fully active, then throw an "InvalidAccessError" DOMException exception.
+    if (!is_fully_active())
+        return throw_completion(WebIDL::InvalidAccessError::create(realm(), "Cannot perform open on a document that isn't fully active."sv));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:507]: Likewise

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.h (hunk 15 above)
Comment (IdanHo) [reply Userland/Libraries/LibWeb/HTML/BrowsingContext.h:129]: Whoops, missed that one, fixed.

**Review by linusg — APPROVED** [review]
LGTM, nice work!

**Pull request merged**